lib/inet-grouping.js

```diff
@@ -1,24 +1,122 @@
+import { base64url } from 'multiformats/bases/base64'
+
 // See https://stackoverflow.com/a/36760050/69868
 const IPV4_REGEX = /^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.){3}(25[0-5]|(2[0-4]|1\d|[1-9]|)\d)$/
 
-export const mapRequestToInetGroup = (/** @type {import('node:http').IncomingMessage} */ req) => {
+/**
+ * @param {import('pg').Client} client
+ * @param {import('node:http').IncomingMessage} req
+ * @param {Date} [now]
+ * @returns {string}
+ */
+export const mapRequestToInetGroup = async (pgClient, req, now = new Date()) => {
+  const subnet = mapRequestToSubnet(req)
+  const group = await mapSubnetToInetGroup(pgClient, subnet, now)
+  return group
+}
+
+/**
+ * @param {import('pg').Client} client
+ * @param {string} subnet
+ * @param {Date} [now]
+ * @returns {Promise<string>}
+ */
+export const mapSubnetToInetGroup = async (pgClient, subnet, now = new Date()) => {
+  const { rows: [found] } = await pgClient.query(
+    'SELECT id FROM inet_groups WHERE subnet = $1',
+    [subnet]
+  )
+  if (found) return found.id
+
+  let remainingRetries = 5
+  for (;; remainingRetries--) {
+    const group = generateUniqueGroupId(now)
+    try {
+      // There is a race condition: between the time we looked up the existing group
+      // and the time we execute this query, a handler for a different request from
+      // the same subnet may have already defined the group for this subnet.
+      // Solution: when insert fails with a conflict in the group id, use that existing group
+      // instead of the value we generated ourselves.
+      //
+      // ON CONFLICT DO UPDATE is needed to tell PG to return the existing id on conflict
+      // ON CONFLICT DO NOTHING would not return the existing id on conflict
+      const { rows: [created] } = await pgClient.query(`
+        INSERT INTO inet_groups (id, subnet, created_at)
+        VALUES ($1, $2, $3)
+        ON CONFLICT (subnet) DO UPDATE SET id = inet_groups.id
+        RETURNING id
+       `, [
+        group,
+        subnet,
+        now
+      ])
+      return created.id
+    } catch (err) {
+      if (err.code === 23505 && err.constraint === 'inet_groups_pkey') {
+        // Retry with a different random id
+        if (remainingRetries > 0) continue
+        // Fail if we exhausted allowed attempts
+        throw new Error('Failed to generate a unique group id')
+      }
+      throw err
+    }
+  }
+}
+
+/**
+ * @param {Date} [now]
+ * @returns {Promise<string>}
+ */
+const generateUniqueGroupId = (now) => {
+  const data = new ArrayBuffer(8)
+
+  const prefix = new Uint8Array(data, 0, 2)
+
+  // To reduce the likelihood of generating the same group values for different subnets,
+  // and also to make it easier to diagnose invalid use of groups (e.g. after they should
+  // have been expired), let's prefix the group with a single byte derived from the number
+  // of hours elapsed since the Unix epoch.
```

Comment:
How is this helping with conflicts? The number of hours elapsed doesn't change for 1h, so all calls in one 30 minutes round see the same hour, right? Or are you worried about conflicts over a long time?

Can we instead use a source of randomness with enough entropy (or long enough group ids), so that we don't need to factor in time or worry about conflicts? Eg something like `sha256(extractSubnet(ip))`?

Reply:
I guess I am too cautious here.

- We need groups to be unique only within each round. It's acceptable to assign the same group to one subnet in round X and a different subnet in round Y. (But I don't think it's a good idea, as it makes troubleshooting more difficult.)
- With 6 random bytes, we have 256^6 different group IDs. There are 256^3 different subnets we are mapping to these groups. With a truly random generator, it would take more than 16 million resets (256^3) before we start recycling the group IDs.
- We are also using `UNIQUE` constraint in our DB to detect duplicate IDs.

The only edge case I can think of is this:
- We happen to reset our mapping table in the middle of a round
- We generate a sequence of random bytes that was already used before the reset (very unlikely, but still has a non-zero probability)
- As a result, two different subnets are mapped to the same group id in this round

The prefix "hours % 251" is handling this edge case if we assume that we will periodically (once a day?) delete group ids older than 24 hours.

> The number of hours elapsed doesn't change for 1h, so all calls in one 30 minutes round see the same hour, right?

Yes, the number does not change for the same round (unless we reset the mapping).

That's why I added the second byte based on milliseconds since epoch.

> Can we instead use a source of randomness with enough entropy

I am using [crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) - I believe that's the most robust random number generator we have available in JavaScript.

---

OK, I think I over-engineered this part. I am going to remove the time but keep using 8 bytes of random data (encoded as 12 characters). That should be good enough to get us started. We can easily tweak this later based on the data we observe.

Reply:
See https://github.com/filecoin-station/spark-api/pull/104/commits/d9f0a40b1501ce41742c0d4feb7c88017f27be6d

Reply:
Sounds good, let's watch for `UNIQUE` constraint violations to see when our randomness isn't good enough?

Reply:
> Sounds good, let's watch for `UNIQUE` constraint violations to see when our randomness isn't good enough?

We are kind of already doing that - when the UNIQUE constraint is violated, we generate a new random id, retrying up to 5 times. If all five attempts fail, then we abort with the following error:

https://github.com/filecoin-station/spark-api/pull/104/files#diff-2ce02f2adf56d846dab8a6c0282e7e21ac77403f11f58c89e08e2d18ff8c15bdR60

```js
throw new Error('Failed to generate a unique group id')
```

This error will be reported to Sentry, so we should be good here.

Reply:
perfect!

Reply:
> This error will be reported to Sentry, so we should be good here.

Ouch, I take that back. We are not logging request-handling errors to Sentry, only unhandled errors crashing the process.

Fixing it here: https://github.com/filecoin-station/spark-api/pull/112